src/lib/frontend/d_cnf.ml

```diff
@@ -786,7 +786,130 @@ let mk_add translate sy ty l =
+    mk_rotate_right n (i - 1) (
+      E.mk_bitv_concat
+        (E.mk_bitv_extract 0 0 t 1)
+        (E.mk_bitv_extract 1 (n - 1) t (n - 1))
+        n
+    )
+
+let mk_bitv_or n t1 t2 =
+  E.mk_term (Sy.Op Sy.BVOr) [t1; t2] (Ty.Tbitv n)
+
+let mk_bitv_and n t1 t2 =
+  E.mk_bvnot n (mk_bitv_or n (E.mk_bvnot n t1) (E.mk_bvnot n t2))
+
+let mk_bitv_xor n t1 t2 =
+  mk_bitv_or n
+    (mk_bitv_and n t1 (E.mk_bvnot n t2))
+    (mk_bitv_and n (E.mk_bvnot n t1) t2)
+
+let mk_bitv_xnor n t1 t2 =
+  mk_bitv_or n
+    (mk_bitv_and n t1 t2)
+    (mk_bitv_and n (E.mk_bvnot n t1) (E.mk_bvnot n t2))
+
+let rec mk_bitv_comp n t1 t2 =
+  if n = 1
+  then mk_bitv_xnor 1 t1 t2
+  else
+    mk_bitv_and 1
+      (mk_bitv_xnor 1
+         (E.mk_bitv_extract (n-1) (n-1) t1 1)
+         (E.mk_bitv_extract (n-1) (n-1) t2 1))
+      (mk_bitv_comp (n-1)
+         (E.mk_bitv_extract 0 (n-2) t1 (n-1))
+         (E.mk_bitv_extract 0 (n-2) t2 (n-1)))
+
+let mk_bv2nat t =
+  E.mk_term (Sy.Op Sy.BV2Nat) [t] Ty.Tint
+
+let mk_nat2bv n t =
+  E.mk_term (Sy.Op (Sy.Nat2BV n)) [t] (Ty.Tbitv n)
+
+let mk_bvneg n x =
+  mk_nat2bv n (
+    E.mk_term (Sy.Op Sy.Minus) [
+      E.mk_term (Sy.Op Sy.Pow) [ E.itwo; E.int (string_of_int n) ] Ty.Tint;
+      mk_bv2nat x
+    ] Ty.Tint
+  )
+
+let mk_bvsign n t =
+  E.mk_bitv_extract (n-1) (n-1) t 1
+
+let bvsign_is_pos t =
+  E.mk_eq ~iff:false t E.bvone
+
+let bvsign_is_neg t =
+  E.mk_eq ~iff:false t E.bvzero
+
+let mk_bvadd n x y =
+  mk_nat2bv n (E.mk_term (Sy.Op Sy.Plus) [mk_bv2nat x; mk_bv2nat y] Ty.Tint)
+
+let mk_bvudiv n x y =
+  (* Assuming that if bv2nat(y) = 0 then bv2nat(x)/bv2nat(y) is
+     undefined and that it makes nat2bv[n](bv2nat(x)/bv2nat(y)) also
+     undefined. *)
+  mk_nat2bv n (E.mk_term (Sy.Op Sy.Div) [mk_bv2nat x; mk_bv2nat y] Ty.Tint)
+
+let mk_bvurem n x y =
+  (* Assuming that if bv2nat(y) = 0 then bv2nat(x)%bv2nat(y) is
+     undefined and that it makes nat2bv[n](bv2nat(x)%bv2nat(y)) also
+     undefined. *)
+  mk_nat2bv n (E.mk_term (Sy.Op Sy.Modulo) [mk_bv2nat x; mk_bv2nat y] Ty.Tint)
+
+let mk_bvlshr n x y =
+  let z = E.mk_term (Sy.Op Sy.Pow) [E.itwo; mk_bv2nat y] Ty.Tint in
+  mk_nat2bv n (E.mk_term (Sy.Op Sy.Div) [mk_bv2nat x; z] Ty.Tint)
+
+let mk_bvult x y =
+  E.mk_builtin ~is_pos:true Sy.LT [ mk_bv2nat x; mk_bv2nat y ]
+
+let mk_bvule x y =
+  E.mk_builtin ~is_pos:true Sy.LE [ mk_bv2nat x; mk_bv2nat y ]
+
+let mk_bvslt n x y =
+  if n = 0 then E.faux else
+    let xsign = mk_bvsign n x in
+    let ysign = mk_bvsign n y in
+    let c1 = E.mk_and (bvsign_is_pos xsign) (bvsign_is_neg ysign) false in
+    let c2 = E.mk_and (E.mk_eq ~iff:false xsign ysign) (mk_bvult x y) false in
+    E.mk_or c1 c2 false
+
+let mk_bvsle n x y =
+  if n = 0 then E.faux else
+    let xsign = mk_bvsign n x in
+    let ysign = mk_bvsign n y in
+    let c1 = E.mk_and (bvsign_is_pos xsign) (bvsign_is_neg ysign) false in
+    let c2 = E.mk_and (E.mk_eq ~iff:false xsign ysign) (mk_bvule x y) false in
+    E.mk_or c1 c2 false
```

Comment:
I understand that these are only used in the Dolmen frontend, but they probably should be in `expr.ml` instead?

Although I am not sure how I feel about eagerly expanding all these; is there a reason these are not constructors with an interpretation? (If the reason is "it is simper this way" it is fine, but we probably should make them constructors later)

Reply:
> I understand that these are only used in the Dolmen frontend, but they probably should be in expr.ml instead?

I'm ok with moving them to `expr.ml`

> If the reason is "it is simper this way"

Yes, that's the only reason. I'll add some constructors and move their simplification to bitv.ml